File: content/atlas/source/includes/avs/index-management/create-index/create-auto-embed-index.cs
using MongoDB.Bson;  
using MongoDB.Bson.Serialization.Conventions;  
using MongoDB.Driver;  
using System;  
using System.Threading;  
  
namespace VectorSearch;  
  
class Program  
{  
    static void Main(string[] args)  
    {  
        // Map title-case class properties to camel-case MongoDB fields  
        var camelCaseConvention = new ConventionPack { new CamelCaseElementNameConvention() };  
        ConventionRegistry.Register("CamelCase", camelCaseConvention, type => true);  
          
        // Connect to your deployment  
        const string mongoConnectionString = "<connectionString>";  
        var client = new MongoClient(mongoConnectionString);  
  
        // Access your database and collection  
        var database = client.GetDatabase("<databaseName>");  
        var collection = database.GetCollection<BsonDocument>("<collectionName>");  
          
        CreateIndex(client, collection, "<indexName>");  
    }  
  
    private static void CreateIndex(MongoClient client, IMongoCollection<BsonDocument> collection, string indexName)  
    {  
        // Create your index model, then create the search index  
        var model = new CreateAutoEmbeddingVectorSearchIndexModel<BsonDocument>(  
            "<fieldToIndex>",      // Field to index  
            indexName,             // Index name  
            "<embeddingModel>"     // Supported Embedding model   
            // Optional: add filter fields as additional parameters if needed  
        );  
  
        var searchIndexView = collection.SearchIndexes;  
        searchIndexView.CreateOne(model);  
        Console.WriteLine($"New search index named {indexName} is building.");  
  
        // Wait for initial sync to complete  
        Console.WriteLine("Polling to check if the index is ready. This may take up to a minute.");  
  
        bool isReady = false;  
        while (!isReady)  
        {  
            var indexes = searchIndexView.List();  
            foreach (var index in indexes.ToEnumerable())  
            {  
                if (index["name"] == indexName)  
                {  
                    isReady = index.Contains("latestDefinition");  
                }  
            }  
  
            if (!isReady)  
            {  
                Thread.Sleep(5000);  
            }  
        }  
          
        Console.WriteLine($"{indexName} is ready for querying.");  
    }  
}  

File: content/vector-search/source/includes/index/autoembed-type/code-snippets/create-index/cpp/create-auto-embed-index.cpp
#include <bsoncxx/builder/basic/document.hpp>
#include <iostream>
#include <mongocxx/client.hpp>
#include <mongocxx/instance.hpp>
#include <mongocxx/search_index_view.hpp>
#include <mongocxx/uri.hpp>
#include <thread>

using bsoncxx::builder::basic::kvp;
using bsoncxx::builder::basic::make_array;
using bsoncxx::builder::basic::make_document;

int main() {
  try {
    mongocxx::instance inst{};

    // Connect to your deployment
    const auto uri = mongocxx::uri{"<connectionString>"};
    mongocxx::client conn{uri};

    // Access your database and collection
    auto db = conn["<databaseName>"];
    auto collection = db["<collectionName>"];

    auto siv = collection.search_indexes();
    std::string name = "<indexName>";

    // Define the index with automated embedding and filter fields
    auto definition = make_document(
        kvp("fields",
            make_array(make_document(
                kvp("type", "autoEmbed"), kvp("modality", "text"),
                kvp("path", "<fieldToIndex>"),
                kvp("model", "<embeddingModel>")))));
    auto model =
        mongocxx::search_index_model(name, definition.view())
            .type("vectorSearch");

    // Create the search index
    siv.create_one(model);
    std::cout << "New search index named " << name << " is building."
              << std::endl;

    // Wait for initial sync to complete
    std::cout << "Polling to check if the index is ready. This may take up to "
                 "a minute."
              << std::endl;
    bool queryable = false;
    while (!queryable) {
      auto indexes = siv.list();
      for (const auto& index : indexes) {
        if (index["name"].get_value() == name) {
          queryable = index["queryable"].get_bool();
        }
      }
      if (!queryable) {
        std::this_thread::sleep_for(std::chrono::seconds(5));
      }
    }
    std::cout << name << " is ready for querying." << std::endl;
  } catch (const std::exception& e) {
    std::cout << "Exception: " << e.what() << std::endl;
  }
  return 0;
}
